authorization.go

```diff
@@ -0,0 +1,204 @@
+	clientOU []string, clientCN string,
+) error {
+	if len(allow.OUs) > 0 {
+		if err := allowedOU(allow.OUs, clientOU); err != nil {
+			return err
+		}
+	}
+	if len(allow.CNs) > 0 {
+		if err := allowedCN(allow.CNs, clientCN); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (allow AllowSpecificOUandCNs) IsAuthorizedForSite(
+	clientOU []string, clientCN, uriSite string,
+) error {
+	// Site resources are not handled separately. Fall back to the behavior
+	// of IsAuthorized
+	return allow.IsAuthorized(clientOU, clientCN)
+}
+
+// RestrictSiteOUs is an instance of AuthorizationCheck that restricts
+// site resource access when the client is in certain OUs.
+// The OUs which this applies to are provided by AppliedOUs.
+// If a client within on of the AppliedOUs requests access to a site resource
+// (i.e. a resource with the `site` URI Parameter), their CN is parsed to obtain their
+// site ID. This site ID must be an exact match of the site ID they are requesting.
+// If the site ID determined from the client's CN does not match the requested resource
+// site ID, the request is denied.
+// If the client's OUs do not match any of `AppliedOUs` the request is allowed.
+// This is intended to be used along with AllowSpecificOUandCNs:
+//    Use AllowSpecificOUandCNs to restrict access to a broad group of OUs
+//    Use RestrictSiteAccess to lock specific OUs into only access their own sites.
+// For example:
+//
+//   opts := certauth.Opts{
+//       AuthorizationChecks: []AuthorizationCheck{
+//           AllowOUs([]string{"admin", "site", "employee"}),
+//           RestrictSiteAccess{
+//               AppliedOUs: []string{"site"},
+//       },
+//   }
+//
+// This example would require *all* requests to present an OU of `admin`, `site`, or
+// `employee` and additionally, requests with an OU of `site` would be prevented from
+// accessing site-specific resources that were not part of their own site.
+type RestrictSiteAccess struct {
+	AppliedOUs []string
+
+	// NOTE(zeal): maybe add a field here to allow the server to choose how the
+	//             site is parsed from the CN?
+}
+
+func (check RestrictSiteAccess) IsAuthorized(clientOU []string, clientCN string) error {
+	// This check only operates on site resources.
+	return nil
+}
+
+func (check RestrictSiteAccess) IsAuthorizedForSite(
+	clientOU []string, clientCN, uriSite string,
+) error {
+	site, _, err := ParseSiteEnvFromCN(clientCN)
+	if err != nil {
+		return err
+	}
+	if site == uriSite {
+		return nil
+	}
+	return fmt.Errorf("Site %q is not authorized to site %q resources", site, uriSite)
+}
+
+// ParseSiteEnvFromCN parses a site id and environment from the provided CN.
+// Also validates that the site ID is a valid UUID.
+// Returns (site, environment, nil) if the clientCN is valid.
+// Returns ("", "", err) if an error occurs.
+func ParseSiteEnvFromCN(clientCN string) (string, string, error) {
+	// Site CN in format of env.site_uuid.domain
```

Comment:
Is that some kind of RFC or 3rd party requirements?

Reply:
it's just what our certs look like

Reply:
Got it.